sway-lib-std/src/storage/storage_vec.sw

```diff
@@ -939,6 +940,61 @@ impl<V> StorageKey<StorageVec<V>> {
             }
         }
     }
+
+    /// Returns a `StorageVecIter` to iterate over this `StorageVec`.
+    ///
+    /// # Returns
+    ///
+    /// * [StorageVecIter<V>] - The struct which can be iterated over.
+    ///
+    /// # Examples
+    ///
+    /// ```sway
+    /// storage {
+    ///     vec: StorageVec<u64> = StorageVec {},
+    /// }
+    ///
+    /// fn foo() {
+    ///     storage.vec.push(5);
+    ///     storage.vec.push(10);
+    ///     storage.vec..push(15);
+    ///
+    ///     // Get the iterator
+    ///     let iter = storage.vec.iter();
+    ///
+    ///     assert(5 == iter.next().unwrap().read());
+    ///     assert(10 == iter.next().unwrap().read());
+    ///     assert(15 == iter.next().unwrap().read());
+    /// }
+    /// ```
+    pub fn iter(self) -> StorageVecIter<V> {
+        StorageVecIter {
+            values: self,
+            index: 0,
+        }
+    }
+}
+
+pub struct StorageVecIter<V> {
+    values: StorageKey<StorageVec<V>>,
+    index: u64,
+}
+
+impl<V> StorageIterator for StorageVecIter<V> {
+    type Item = StorageKey<V>;
+    #[storage(read)]
+    fn next(ref mut self) -> Option<Self::Item> {
+        if self.index >= self.values.len() {
```

Comment:
This seriously increases the gas cost of an iteration, compared to a classical approach with the `while` loop and an index.

In the classical approach, the `len()` will be called, only once, before entering the loop. In this case, it is called for every iteration step.

I have strong concerns in providing a convenient way to iterate all elements, that can potentially double the gas cost of the iteration.

A simple solution can be to have a `len` field in the `StorageVecIter` and initialize it to the vector length in the `iter()`. This is of course fragile if the vector gets changed during the iteration. But changing the collection while iterating over it is a logical error anyhow. Unfortunately, unlike Rust, Sway cannot forbid it on the language level.